cpp/src/copying/contiguous_split.cu

```diff
@@ -14,466 +14,1105 @@
+    output.first->offset_stack_pos = output.second;
+    output.second += offset_depth;
+    output.first->parent_offsets_index = parent_offset_index;
+    output.first->is_validity          = false;
+    output.first++;
+
+    return output;
+  }
+
+ private:
+  src_buf_output maybe_add_null_buffer(column_view const& col,
+                                       src_buf_output output,
+                                       int parent_offset_index,
+                                       int offset_depth)
+  {
+    if (!col.nullable()) { return output; }
+    output.first->type             = type_id::INT32;
+    output.first->offsets          = nullptr;
+    output.first->offset_stack_pos = output.second;
+    output.second += offset_depth;
+    output.first->parent_offsets_index = parent_offset_index;
+    output.first->is_validity          = true;
+    output.first++;
+
+    return output;
+  }
+};
+
+template <typename InputIter>
+src_buf_output setup_source_buf_info(InputIter begin,
+                                     InputIter end,
+                                     src_buf_info* head,
+                                     src_buf_output output,
+                                     int parent_offset_index,
+                                     int offset_depth)
+{
   std::for_each(
-    t.begin(), t.end(), [&out_cols, &buf, &column_index, &split_info](cudf::column_view const& c) {
-      cudf::type_dispatcher(
-        c.type(), column_copy_functor{}, c, split_info[column_index], buf, out_cols);
-      column_index++;
+    begin, end, [head, &output, parent_offset_index, offset_depth](column_view const& col) {
+      output = cudf::type_dispatcher(
+        col.type(), buf_info_functor{}, col, head, output, parent_offset_index, offset_depth);
     });
+  return output;
+}
+
+/**
+ * @brief Given a set of input columns and computed split buffers, produce
+ * output columns.
+ *
+ * After performing the split we are left with 1 large buffer per incoming split
+ * partition.  We need to traverse this buffer and distribute the individual
+ * subpieces that represent individual columns and children to produce the final
+ * output columns.
+ *
+ * This function is called recursively in the case of nested types.
+ *
+ * @param begin Beginning of input columns
+ * @param end End of input columns
+ * @param dst Output vector of column views to produce
+ * @param partition_index Partition index (which section of the split we are working with)
+ * @param buf_index Absolute split output buffer index
+ * @param h_dst_bufs Array of destination buffers (indexed by partition_index)
+ * @param h_dst_buf_info Array of offsets into destination buffers (indexed by buf_index)
+ *
+ * @returns new working buffer index after processing this range of input columns
+ */
+template <typename InputIter>
+size_type build_output_columns(InputIter begin,
+                               InputIter end,
+                               std::vector<column_view>& dst,
```

Comment:
Output parameter.

Reply:
Can you be more specific? Do you want it 
- returning a column_view and turning the index parameter into a reference
- returning a pair containing a column_view and the index parameter